### DISCUSSION TITLE
UTC Ham Clock

### DISCUSSION THREAD

**Discussion by Ken-Kauffmann** [arnaud-j4k08/PicoClockGreenEasy#5] [Ideas]
Amateur Radio operators use UTC dates and times for logging, propagation reports, and digital communications.  Many hams keep a separate clock in their shack, set to UTC time. 
The PicoClockGreenEasy can easily be set up as a UTC Hamclock by setting the UTC_OFFSET to 0, Time Format to 24H, and DST_LOCATION to Unknown

As a Ham Radio operator, a "nice to have" function would be to have the PicoClockGreenEasy scroll between Local Time and UTC time.  Maybe with horizontal scrolling that reads "Local Time: 10 : 50 Date: 12:25" periodicallly changing to  "UTC Time: 15:50 Date: 12:25" 

Of course, I could just have two PicoClockGreenEasys, if I was agile enough.  :-)  Do software engineers still use Agile?

**Comment by Ken-Kauffmann** [comment]
I am thinking that a HAMClock could be added fairly easily.  ( I will attempt it myself, and learn much more about your implementation.)

HAMCLOCK would be an additional entry in the Options menu, along with the stopwatch, and countdown timer.

When selected, this option would operate similarly to the current autoscroll which shows Time, Date, and Temperature.  In my case, there would be 4 functions, which repeat in a loop.

1)  Local Time in 12H format.  The presence of the AM and PM indicators would signify to the user that this is the local time.  I would try to implement this by calling the existing routines, and would allow for update of the time.  

2) Local Date in MM:DD, or DD:MM format.   The format could be triggered by the existing DST_LOCATION variable.  I would try to implement this by calling the existing routines, and would allow for update of the date. 

3) UTC Time in 24H format.  The absence of the AM and PM indicators would signify to the user that this is the UTC time.  I would try to implement this by replicating the existing routine, but showing the UTC time.  This routine would not allow updates to the time if UTC_OFFSET has been set, and UTC_OFFSET will be saved to a different variable, m_utc-offset, for calculation off of  Local time.  The routine would allow updates if UTC_OFFSET has not been set.  When the user sets the UTC time, rather than saving the new time, I would calculate a m_utc-offset and use it for future display purposes.  

4) UTC Date in MM:DD or DD:MM format.  The format could be triggered by the existing DST_LOCATION variable.  I would implement this by replicating the existing Date routines, but using the UTC time rather than Local time in the date derivation.  In addition, if the UTC date was different from the Local date,  I would cause the date to blink.   The user would not be able to update the UTC Date. 

The three buttons would act the same as normal.  Buttons 2 and 3 would act as up or down scrolling, or in field updates.  Button 1 would initiate the update sequence, where updates are allowed. 

I would imagine that this would still work if Auto Scrolling was turned off, with the up and down buttons moving between the lines. 

If I must replicate all of the 4 routines, I might implement the buttons a little differently.  Scrolling would basically only go one-way: From Local Time to Local Date to UTC Time to UTC Date back to Local Time.   Button 1 and Button 3 would act as before.  Button 2 would allow the user to toggle through the 3 Time formats, on the two time displays.  Local Time would be restricted to 12h formats, and UTC Time would be restricted to 24 hour formats.  

The world does not really need another Ham Clock.  But, you have already done all of the heavy lifting.  I will begin by studying and annotating ClockUI.  I see that the functions are added to a root menu, and the magic happens later.  I  suspect that I will have to replicate much of the workings of ClockUI, in a new HamClockUi.   In any case, it should be fun, and a good learning experience.

I will avoid pestering you as much as possible.  I need to remember that I am the retired one, and that you still have a day job.

Any thoughts regarding my proposed implementation?

**Comment by Ken-Kauffmann** [comment]
I have done some investigation, and I think that I have found most of the bits that I need.  First off, I think that the HamClock needs to be at the same level as the Stopwatch, or Countdown timer.  Trying to launch it from Options.cpp isn't going to fly.  I think that I can use Time::HourMin, and Date as is.  I will try to put them into *hamclock* submenu.  I think that I can add a Time::UTCHourMin in Time.cpp and force it into 24H mode, and create a *hamclock* submenu entry for that.   I will need UTC time for this, and don't know that I want to calculate it each time.  I am looking at an alternative, ( see below).  I think that the easiest way to implement the 4th submenu is to replicate Date.cpp as UTCdate,cpp, and make my changes,  eliminating the modification functions.  I would put it in the *hamclock* submenu in the same manner as Date above.    As I understand submenus, I think that I will be getting a 5th EXIT function automatically.  It would be nice if AutoScroll will loop past it, and back to the Local time submenu.  This might happen magically if isAvailable is being set for the existing functions.  
As described above, I need the UTC time in seconds, and a UTC tm structure.  My understanding is that Clock,cpp is constantly updating a m_tm and m_time.  They are also set from the RTC, and through the NTP sync.  The various renderings in Date and Time::HourMin use a clock().get() function to obtain the hours, minutes, months and days from the m_tm structure.  My understanding it that these  are Local times, without DST applied.  Rather than calculating an UTC tm structure each time I need to display, I am considering creating m_UTCtm, and m_UTCtime.  There are not that many places in Clock.cpp that set m_tm and m_time, and setting m_UTCtm and m_UTCtime at the same moment would not be difficult.  In most cases, it is copying the statement, and removing the UTC_OFFSET component.  

Please let me know if I am on the wrong track.  Thanks, Ken

**Comment by arnaud-j4k08** [comment]
>I have done some investigation, and I think that I have found most of the bits that I need. First off, I think that the HamClock needs to be at the same level as the Stopwatch, or Countdown timer. Trying to launch it from Options.cpp isn't going to fly.

Yes. I understood that your initial idea was to integrate a kind of additional application to PicoClockGreenEasy, thus a an additional UI class next to ClockUi. This is of course technically feasible, but a lot more work.

>I think that I can use Time::HourMin, and Date as is. I will try to put them into hamclock submenu. I think that I can add a Time::UTCHourMin in Time.cpp and force it into 24H mode, and create a hamclock submenu entry for that.

You can indeed extend the Time class and add a new style to the Time::Style enum. You can then create a additional instance of it and put it in another menu. As you probably already found out, https://github.com/arnaud-j4k08/PicoClockGreenEasy/blob/000db551cdb0898c8f03b51414001c6e4c6393e1/src/ClockUi.cpp#L54 is the location that builds the tree of functions described in https://github.com/arnaud-j4k08/PicoClockGreenEasy/blob/000db551cdb0898c8f03b51414001c6e4c6393e1/README.md?plain=1#L106. 

>I will need UTC time for this, and don't know that I want to calculate it each time. I am looking at an alternative, ( see below). I think that the easiest way to implement the 4th submenu is to replicate Date.cpp as UTCdate,cpp, and make my changes, eliminating the modification functions. I would put it in the hamclock submenu in the same manner as Date above. 

You can either replicate Date.cpp or extend it with a flag to function in UTC mode, and instance it a second time in ClockUi to put in the hamclock submenu.

>As I understand submenus, I think that I will be getting a 5th EXIT function automatically. 

Yes, submenus get an "Exit" function automatically. This is implemented in https://github.com/arnaud-j4k08/PicoClockGreenEasy/blob/000db551cdb0898c8f03b51414001c6e4c6393e1/src/Functions/Submenu.cpp#L7.

>It would be nice if AutoScroll will loop past it, and back to the Local time submenu. 

I am not sure it will work as you expect. The autoscroll switching is done at https://github.com/arnaud-j4k08/PicoClockGreenEasy/blob/000db551cdb0898c8f03b51414001c6e4c6393e1/src/ClockUi.cpp#L163 to move between 3 functions at the 0th, 20th and 40th second of every meeting. This may require some adaption.

>This might happen magically if isAvailable is being set for the existing functions.

isAvailable returns true by default for most functions. It is used to e.g. make the "Wifi status" function disappear if not Wifi is configured at all.

>As described above, I need the UTC time in seconds, and a UTC tm structure. My understanding is that Clock,cpp is constantly updating a m_tm and m_time. They are also set from the RTC, and through the NTP sync. The various renderings in Date and Time::HourMin use a clock().get() function to obtain the hours, minutes, months and days from the m_tm structure. My understanding it that these are Local times, without DST applied. 

Yes, that's how it works. All these times are local times. However, m_time and m_tm differ regarding DST consideration:

```
    time_t m_time = 0; // Current time as unix time, local (not UTC), not considering DST
    tm m_tm = {}; // Current time as tm, considering DST

```
Thus, clock().get() also returns time with DST applied.

>Rather than calculating an UTC tm structure each time I need to display, I am considering creating m_UTCtm, and m_UTCtime. There are not that many places in Clock.cpp that set m_tm and m_time, and setting m_UTCtm and m_UTCtime at the same moment would not be difficult. In most cases, it is copying the statement, and removing the UTC_OFFSET component.

You will probably don't need any m_UTCtime member variable, but only a local variable, as the cost of performing the UTC_OFFSET substraction on every second is negligeable. Then, the Clock::setTmFromTime method will need to contain something like:
```
    time_t utcTime = m_time - UTC_OFFSET * 3600;
    m_UTCtm = *localtime(&utcTime);
```

>Please let me know if I am on the wrong track. Thanks, Ken

No, it looks good. :-)

**Comment by Ken-Kauffmann** [comment]
Thank you for your quick, and detailed response.  You have given me the confidence to proceed further.  As I am a beginner, I will start slowly, and piecemeal.  I think that I will start by creating the *hamclock* submenu with the two "known good" routines of Time::HourMin, and Date.  If I can get that to work, I might put the UTC changes into the existing DATE, using the two statements you provided above.  If that works, then either create a UTCDate.cpp, or figure out a way to extend Date.cpp with a flag.  In my current ignorance, it is probably easier for me to create the UTCDate.cpp, as I just need to copy statements everywhere the current version appears.  I really am a bit of a monkey at this stage.  

In any case, that is my plan.  Start from a working version, make a small change, regroup if necessary.  

I am doing this for fun, and I am not in a hurry.  We live in a 3 generation household, and the grandkids are home from school until after the New Year.  All that is to say that I may not make too much progress for a while, and will finally give you a break from answering my noob questions.  Thanks again for your help, and for creating such a fantastic piece of code.

**Reply by arnaud-j4k08** [reply]
That sounds like a good plan. As you're not in a hurry and you're doing this just for fun (like me), nothing can go wrong. Thank you again for your compliments. I have put this project on github in case it's useful to someone, and this has definitely worked.

I wish you and your family a Merry Christmas!

**Comment by Ken-Kauffmann** [comment]
Success!  The actual code changes were fairly small, after I understood how the process worked.  I added a new style to Time, and modified my version of Date to use  styles of Local and UTC.  I created a hamclockSubmenu by copying the form for Alarms, and doing addFunctions for my UTC subfunctions.  The manual scrolling of the submenu works great, and the Exit works as intended.  I modified my versions of Time and Date to display the C symbol when displaying a UTC date or time.  I have even modified the existing AutoScroll block to display my UTC dates and times in sequence with the local date, time, and temperature. 
It is a testament to your work, that I was able to make these changes easily, and perhaps more importantly, that I was able to understand how to make the changes.  Thanks, Ken

**Reply by arnaud-j4k08** [reply]
Thank you. I am glad you were able to extend the code to have the feature you designed, without much difficulty.

I am preparing a new version and currently adding better synchronization features, as discussed in this other thread: https://github.com/arnaud-j4k08/PicoClockGreenEasy/discussions/4
